### PR TITLE
CreateCompilation default option to detect 'unsafe' keyword

### PR DESCRIPTION
**Pull request by jcouv**
Small convenience: When options aren't specified, `CreateCompilation` can look for `unsafe` in the syntax to determine what options I need.

### CODE REVIEW

File: src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs [hunk 1]
```diff
@@ -1250,12 +1250,19 @@ public static SyntaxTree ParseWithRoundTripCheck(string text, CSharpParseOptions
             return compilation;
         }
 
-        private static CSharpCompilationOptions CheckForTopLevelStatements(SyntaxTree[] syntaxTrees)
+        private static CSharpCompilationOptions CheckForTopLevelStatementsAndUnsafe(SyntaxTree[] syntaxTrees)
```
Comment (333fred) [inline src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:1253]: At this point is it worth making the check a syntax walker that returns a flag enum? Yesterday it's tls, today it's tls and unsafe, tomorrow it's another new thing?

Comment (jcouv) [reply src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:1253]: Returning options directly is fine, what's the benefit of adding an extra layer of indirection?

Comment (333fred) [reply src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:1253]: Avoid an extra syntax walk on nearly every syntax tree in nearly every compiler test, and avoiding allocations of new compilation options for unsafe tests.

Comment (jcouv) [reply src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:1253]: I see. I realized what you meant this afternoon. I was slow ;-) Sure!

File: src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs [hunk 2]
```diff
@@ -1250,18 +1250,50 @@ public static SyntaxTree ParseWithRoundTripCheck(string text, CSharpParseOptions
...
+
+            public override void Visit(SyntaxNode node)
             {
-                return tree.GetRoot().DescendantTokens().Any(t => t.IsKind(SyntaxKind.UnsafeKeyword));
+                if (node.Kind() is SyntaxKind.GlobalStatement)
+                {
+                    _foundTopLevelStatement = true;
+                }
+
+                base.Visit(node);
```
Comment (333fred) [inline src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:1288]: Consider returning early after `_foundTopLevelStatement` and `_foundUnsafe` are true.

File: src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs [hunk 3]
```diff
@@ -1250,18 +1250,50 @@ public static SyntaxTree ParseWithRoundTripCheck(string text, CSharpParseOptions
...
+                    syntaxWalker.Visit(tree.GetRoot());
+                    if (syntaxWalker._foundTopLevelStatement && syntaxWalker._foundUnsafe)
+                    {
+                        continue;
+                    }
+                }
 
-            return options;
+                var options = syntaxWalker._foundTopLevelStatement ? TestOptions.ReleaseExe : TestOptions.ReleaseDll;
+                options = options.WithAllowUnsafe(syntaxWalker._foundUnsafe);
```
Comment (333fred) [inline src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:1276]: Consider just doing a switch expression over the booleans and using the singleton `UnsafeReleaseExe`/`UnsafeReleaseDll` options on `TestOptions`.

**Comment by AlekseyTs** [comment]
> Small convenience: When options aren't specified, CreateCompilation can look for unsafe in the syntax to determine what options I need.

I do not think we should have this convenience inside a general helper. In my opinion this behavior should not be default, i.e. must be explicitly requested. Alternatively
- a helpers with specific name can be added. 
- a helper to infer options from code can be added.
- ...

The bottom line, the fact that we take advantage of convenience should be clearly visible in the source code of a test that takes advantage of it. That, BTW, will eliminate the need to modify any existing tests.

**Comment by AlekseyTs** [comment]
I understand that the change might feel like an "innocent" extension of what we do for top level statements. However, I think that an implicit inference of something more than DLL vs. EXE is a step too far.

**Comment by jcouv** [comment]
Closing per discussion with the team. Thanks!

**Pull request closed without merging**